api/utils/keys/hardwarekey/pincache.go

```diff
@@ -0,0 +1,113 @@
+// Copyright 2025 Gravitational, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package hardwarekey
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/gravitational/trace"
+	"github.com/jonboulle/clockwork"
+)
+
+// PINCache is a PIN cache that supports consumers with varying required TTLs.
+type PINCache struct {
+	Clock clockwork.Clock
+
+	mu sync.Mutex
+	// pin is the cached PIN.
+	pin string
+	// pinSetAt is the time when the cached PIN was set. Used to determine whether
+	// the PIN should be considered expired for a specific caller's provided TTL.
+	pinSetAt time.Time
+	// pinExpiry is the expiration time of the cached PIN.
+	pinExpiry time.Time
+}
+
+// NewPINCache returns a new PINCache.
+func NewPINCache() *PINCache {
+	return &PINCache{
+		Clock: clockwork.NewRealClock(),
+	}
+}
+
+// GetPIN retrieves the cached PIN. If the PIN was cached before by an amount of
+// time equal to the provided TTL, the PIN will not be returned.
+func (p *PINCache) GetPIN(ttl time.Duration) string {
+	p.mu.Lock()
+	defer p.mu.Unlock()
+
+	if p.pin == "" {
+		return ""
+	}
+
+	// Check if the PIN cache is expired. If it is, wipe it.
+	if p.Clock.Now().After(p.pinExpiry) {
+		p.pin = ""
+		p.pinExpiry = time.Time{}
+		p.pinSetAt = time.Time{}
+		return ""
+	}
+
+	// The PIN is cached, but does not satisfy the provided TTL of the request.
+	// e.g. it has been alive for 8 minutes, but the provided TTL is 5 minutes.
+	// For the purposes of this request, the pin should be considered expired.
+	expiredForRequest := p.pinSetAt.Add(ttl)
+	if !p.Clock.Now().Before(expiredForRequest) {
+		return ""
+	}
+
+	return p.pin
+}
+
+// SetPIN sets the given PIN in the cache with the given TTL. If the PIN
+// is already cached, the existing expiration is only updated if the given
+// TTL would exceed that expiration.
+func (p *PINCache) SetPIN(pin string, ttl time.Duration) {
+	if ttl == 0 {
```

Comment:
Should the cached pin be updated in this case?

Reply:
`ttl=0` should be a no-op for both `GetPIN` and `SetPIN`. But it's awkward to even call either of them. I'll move this check to `PromptOrGetPIN`:

```
func (p *pinCache) PromptOrGetPIN(ctx context.Context, prompt hardwarekey.Prompt, requirement hardwarekey.PINPromptRequirement, keyInfo hardwarekey.ContextualKeyInfo, pinCacheTTL time.Duration) (string, error) {
	// If the provided ttl is 0, it doesn't support caching, so we just prompt.
	if pinCacheTTL == 0 {
		return prompt.AskPIN(ctx, requirement, keyInfo)
	}
	...
```